src/PayerRegistry.sol

```diff
@@ -0,0 +1,530 @@
+        PayerRegistryStorage storage $ = _getPayerRegistryStorage();
+        Payer storage payer_ = $.payers[msg.sender];
+        uint96 pendingWithdrawal_ = payer_.pendingWithdrawal;
+
+        require(pendingWithdrawal_ > 0, NoPendingWithdrawal());
+        require(payer_.balance >= 0, PayerInDebt());
+
+        // slither-disable-next-line timestamp
+        require(
+            block.timestamp >= payer_.withdrawableTimestamp,
+            WithdrawalNotReady(uint32(block.timestamp), payer_.withdrawableTimestamp)
+        );
+
+        delete payer_.pendingWithdrawal;
+        delete payer_.withdrawableTimestamp;
+
+        $.totalDeposits -= _toInt104(pendingWithdrawal_);
+
+        emit WithdrawalFinalized(msg.sender);
+
+        require(ERC20Helper.transfer(token, recipient_, pendingWithdrawal_), ERC20TransferFailed());
+    }
+
+    /// @inheritdoc IPayerRegistry
+    function settleUsage(address[] calldata payers_, uint96[] calldata fees_) external onlySettler whenNotPaused {
+        require(payers_.length == fees_.length, ArrayLengthMismatch());
+
+        PayerRegistryStorage storage $ = _getPayerRegistryStorage();
+        int104 totalDeposits_ = $.totalDeposits;
+        uint96 totalDebt_ = $.totalDebt;
+        int104 minActiveBalance_ = minActiveBalance();
+
+        for (uint256 index_; index_ < payers_.length; ++index_) {
+            address payer_ = payers_[index_];
+            uint96 fee_ = fees_[index_];
+
+            emit UsageSettled(payer_, fee_);
+
+            totalDeposits_ -= _toInt104(fee_);
+            totalDebt_ += _decreaseBalance(payer_, fee_, minActiveBalance_);
+        }
+
+        $.totalDeposits = totalDeposits_;
+        $.totalDebt = totalDebt_;
+
+        // The excess token in the contract that is not withdrawable by payers can be sent to the fee distributor.
+        uint96 totalWithdrawable_ = _getTotalWithdrawable(totalDeposits_, totalDebt_);
+        uint96 tokenBalance_ = uint96(IERC20Like(token).balanceOf(address(this)));
+        uint96 excess_ = tokenBalance_ > totalWithdrawable_ ? tokenBalance_ - totalWithdrawable_ : 0;
+
+        // slither-disable-next-line incorrect-equality
+        if (excess_ == 0) return;
+
+        emit FeesTransferred(excess_);
+
+        require(ERC20Helper.transfer(token, $.feeDistributor, excess_), ERC20TransferFailed());
+    }
+
+    /// @inheritdoc IPayerRegistry
+    function updatePayerStatuses(address[] calldata payers_) external whenNotPaused {
+        PayerRegistryStorage storage $ = _getPayerRegistryStorage();
+        int104 minActiveBalance_ = minActiveBalance();
+
+        for (uint256 index_; index_ < payers_.length; ++index_) {
+            address payer_ = payers_[index_];
+
+            _activatePayerIfNecessary(payer_, $.payers[payer_].balance, minActiveBalance_);
+            _deactivatePayerIfNecessary(payer_, $.payers[payer_].balance, minActiveBalance_);
+        }
+    }
```

Comment:
Who and when would call this function? Seems redundant as the contract alone should be capable of keeping its balances and activation/deactivations.

Reply:
> Who and when would call this function? 
Anyone with a vested interest in toggling the subset of `payers_` to the correct status, be it the payer themself, a node operator, a watchdog, a dev.

> Seems redundant as the contract alone should be capable of keeping its balances and activation/deactivations.
The natspec comment in the interface addresses this
`@notice Should be called after changing the minimum active balance.`
The `setMinActiveBalance` cannot be tasked with looping through all payers to update their inclusion in the set, so this convenience function is necessary to satisfy the active payers convenience functionality such that the set eventually settles to be correct.

Reply:
I was msotly thinking about the (de)activation logic, which is essentially what we do by calling this func. It doesn't seem it's needed anymore.

Reply:
Periphery removed in latest commit, as per agreement on call.